File: SWEagent/src/rules/project-overview.ts
/**
 * Project overview and structure
 */

import { ProjectStructure, EntryPoint, ClassInfo, ExecutionEnvironment, ToolsInfo, Inspector } from './types';

/**
 * Main entry points for the SWE-agent application
 */
export const ENTRY_POINTS: EntryPoint[] = [
  {
    path: 'sweagent/run/run_single.py',
    description: 'Main entry point for running single agent instances',
    type: 'single',
  },
  {
    path: 'sweagent/run/run_batch.py',
    description: 'Entry point for batch processing, used for benchmarking',
    type: 'batch',
  },
];

/**
 * Main agent class information
 */
export const MAIN_AGENT_CLASS: ClassInfo = {
  name: 'Agent',
  path: 'sweagent/agent/agents.py',
  description: 'Main class that governs agent behavior',
};

/**
 * SWE environment class information
 */
export const SWE_ENV_CLASS: ClassInfo = {
  name: 'SWEEnv',
  path: 'sweagent/environment/swe_env.py',
  description: 'Class that interfaces with the SWE-ReX project to interface with sandboxed docker containers',
};

/**
 * Execution environment configuration
 */
export const EXECUTION_ENVIRONMENT: ExecutionEnvironment = {
  type: 'docker',
  description: 'AI agent proposes actions that are executed in sandboxed docker containers',
  interfaceClass: SWE_ENV_CLASS,
  interfaceProject: 'SWE-ReX',
};

/**
 * Tools configuration
 */
export const TOOLS_INFO: ToolsInfo = {
  location: 'tools/',
  organization: 'bundles',
  deployment: 'Copied to sandboxed container and made available in $PATH variable',
};

/**
 * Inspector tools for trajectory analysis
 */
export const INSPECTORS: Inspector[] = [
  {
    name: 'CLI Inspector',
    path: 'inspector_cli.py',
    type: 'cli',
    description: 'Command line interface for inspecting agent trajectories',
  },
  {
    name: 'Web Inspector',
    path: 'sweagent/inspector/server.py',
    type: 'web',
    description: 'Web interface for inspecting agent trajectories',
  },
];

/**
 * Complete project structure
 */
export const PROJECT_STRUCTURE: ProjectStructure = {
  mainEntryPoints: ENTRY_POINTS,
  mainClass: MAIN_AGENT_CLASS,
  executionEnvironment: EXECUTION_ENVIRONMENT,
  tools: TOOLS_INFO,
  inspectors: INSPECTORS,
};

/**
 * Project overview text
 */
export const PROJECT_OVERVIEW = `SWE-agent implements an AI software engineering agent that uses language models to fix github issues.`;

/**
 * Get project component by path
 */
export function getComponentByPath(path: string): { component: string; description: string } | null {
  // Check entry points
  const entryPoint = ENTRY_POINTS.find((ep) => ep.path === path);
  if (entryPoint) {
    return { component: 'entry-point', description: entryPoint.description };
  }

  // Check main classes
  if (path === MAIN_AGENT_CLASS.path) {
    return { component: 'main-agent', description: MAIN_AGENT_CLASS.description };
  }

  if (path === SWE_ENV_CLASS.path) {
    return { component: 'environment', description: SWE_ENV_CLASS.description };
  }

  // Check inspectors
  const inspector = INSPECTORS.find((i) => i.path === path);
  if (inspector) {
    return { component: 'inspector', description: inspector.description };
  }

  // Check if it's a tool
  if (path.startsWith('tools/')) {
    return { component: 'tool', description: 'Tool bundle for agent actions' };
  }

  return null;
}

/**
 * Get all Python modules in the project structure
 */
export function getPythonModules(): string[] {
  const modules: string[] = [];

  // Add entry points
  modules.push(...ENTRY_POINTS.map((ep) => ep.path));

  // Add main classes
  modules.push(MAIN_AGENT_CLASS.path);
  modules.push(SWE_ENV_CLASS.path);

  // Add inspectors
  modules.push(...INSPECTORS.map((i) => i.path));

  return modules;
}

/**
 * TypeScript equivalent paths (for conversion tracking)
 */
export const TYPESCRIPT_EQUIVALENTS: Record<string, string> = {
  'sweagent/run/run_single.py': 'src/run/run-single.ts',
  'sweagent/run/run_batch.py': 'src/run/run-batch.ts',
  'sweagent/agent/agents.py': 'src/agent/agents.ts',
  'sweagent/environment/swe_env.py': 'src/environment/swe-env.ts',
  'inspector_cli.py': 'src/inspector/cli.ts',
  'sweagent/inspector/server.py': 'src/inspector/server.ts',
};

File: cpp_merge/packages/integration/SWEagent/src/rules/project-overview.cpp
#include "project-overview.h"

any getComponentByPath(string path)
{
    auto entryPoint = ENTRY_POINTS->find([=](auto ep) mutable
    {
        return ep->path == path;
    }
    );
    if (entryPoint) {
        return object{
            object::pair{std:("component"), std:("entry-point")}, 
            object::pair{std:("description"), entryPoint->description}
        };
    }
    if (path == MAIN_AGENT_CLASS->path) {
        return object{
            object::pair{std:("component"), std:("main-agent")}, 
            object::pair{std:("description"), MAIN_AGENT_CLASS->description}
        };
    }
    if (path == SWE_ENV_CLASS->path) {
        return object{
            object::pair{std:("component"), std:("environment")}, 
            object::pair{std:("description"), SWE_ENV_CLASS->description}
        };
    }
    auto inspector = INSPECTORS->find([=](auto i) mutable
    {
        return i->path == path;
    }
    );
    if (inspector) {
        return object{
            object::pair{std:("component"), std:("inspector")}, 
            object::pair{std:("description"), inspector->description}
        };
    }
    if (path->startsWith(std:("tools/"))) {
        return object{
            object::pair{std:("component"), std:("tool")}, 
            object::pair{std:("description"), std:("Tool bundle for agent actions")}
        };
    }
    return nullptr;
};


array<string> getPythonModules()
{
    auto modules = array<string>();
    modules->push(const_(ENTRY_POINTS->map([=](auto ep) mutable
    {
        return ep->path;
    }
    ))[0]);
    modules->push(MAIN_AGENT_CLASS->path);
    modules->push(SWE_ENV_CLASS->path);
    modules->push(const_(INSPECTORS->map([=](auto i) mutable
    {
        return i->path;
    }
    ))[0]);
    return modules;
};


array<std::shared_ptr<EntryPoint>> ENTRY_POINTS = array<std::shared_ptr<EntryPoint>>{ object{
    object::pair{std:("path"), std:("sweagent/run/run_single.py")}, 
    object::pair{std:("description"), std:("Main entry point for running single agent instances")}, 
    object::pair{std:("type"), std:("single")}
}, object{
    object::pair{std:("path"), std:("sweagent/run/run_batch.py")}, 
    object::pair{std:("description"), std:("Entry point for batch processing, used for benchmarking")}, 
    object::pair{std:("type"), std:("batch")}
} };
std::shared_ptr<ClassInfo> MAIN_AGENT_CLASS = object{
    object::pair{std:("name"), std:("Agent")}, 
    object::pair{std:("path"), std:("sweagent/agent/agents.py")}, 
    object::pair{std:("description"), std:("Main class that governs agent behavior")}
};
std::shared_ptr<ClassInfo> SWE_ENV_CLASS = object{
    object::pair{std:("name"), std:("SWEEnv")}, 
    object::pair{std:("path"), std:("sweagent/environment/swe_env.py")}, 
    object::pair{std:("description"), std:("Class that interfaces with the SWE-ReX project to interface with sandboxed docker containers")}
};
std::shared_ptr<ExecutionEnvironment> EXECUTION_ENVIRONMENT = object{
    object::pair{std:("type"), std:("docker")}, 
    object::pair{std:("description"), std:("AI agent proposes actions that are executed in sandboxed docker containers")}, 
    object::pair{std:("interfaceClass"), SWE_ENV_CLASS}, 
    object::pair{std:("interfaceProject"), std:("SWE-ReX")}
};
std::shared_ptr<ToolsInfo> TOOLS_INFO = object{
    object::pair{std:("location"), std:("tools/")}, 
    object::pair{std:("organization"), std:("bundles")}, 
    object::pair{std:("deployment"), std:("Copied to sandboxed container and made available in $PATH variable")}
};
array<std::shared_ptr<Inspector>> INSPECTORS = array<std::shared_ptr<Inspector>>{ object{
    object::pair{std:("name"), std:("CLI Inspector")}, 
    object::pair{std:("path"), std:("inspector_cli.py")}, 
    object::pair{std:("type"), std:("cli")}, 
    object::pair{std:("description"), std:("Command line interface for inspecting agent trajectories")}
}, object{
    object::pair{std:("name"), std:("Web Inspector")}, 
    object::pair{std:("path"), std:("sweagent/inspector/server.py")}, 
    object::pair{std:("type"), std:("web")}, 
    object::pair{std:("description"), std:("Web interface for inspecting agent trajectories")}
} };
std::shared_ptr<ProjectStructure> PROJECT_STRUCTURE = object{
    object::pair{std:("mainEntryPoints"), ENTRY_POINTS}, 
    object::pair{std:("mainClass"), MAIN_AGENT_CLASS}, 
    object::pair{std:("executionEnvironment"), EXECUTION_ENVIRONMENT}, 
    object::pair{std:("tools"), TOOLS_INFO}, 
    object::pair{std:("inspectors"), INSPECTORS}
};
string PROJECT_OVERVIEW = std:("SWE-agent implements an AI software engineering agent that uses language models to fix github issues.");
Record<string, string> TYPESCRIPT_EQUIVALENTS = object{
    object::pair{std:("sweagent/run/run_single.py"), std:("src/run/run-single.ts")}, 
    object::pair{std:("sweagent/run/run_batch.py"), std:("src/run/run-batch.ts")}, 
    object::pair{std:("sweagent/agent/agents.py"), std:("src/agent/agents.ts")}, 
    object::pair{std:("sweagent/environment/swe_env.py"), std:("src/environment/swe-env.ts")}, 
    object::pair{std:("inspector_cli.py"), std:("src/inspector/cli.ts")}, 
    object::pair{std:("sweagent/inspector/server.py"), std:("src/inspector/server.ts")}
};

void Main(void)
{
}

MAIN
